### PR TITLE
Allow dollar signs for math in the documentation

### PR DESCRIPTION
**Pull request by asmeurer**
#### References to other Issues or PRs

This is the first step to fixing https://github.com/sympy/sympy/issues/13519.

I have modified the docstrings in bessel.py as a demonstration. I do not plan on modifying other docstrings at this time. 

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by asmeurer** [comment]
I haven't added support for display math `$$math$$`. For now you will need to continue to use `.. math::` for it. It shouldn't be hard to add support for it if it is desired.

**Comment by asmeurer** [comment]
This has some false positives in the latex() docstring. The thing I wrote in the commit message was wrong. I think the simplest fix here is to disable the extension for that docstring. Ideally we would be able to parse $ just like a real Sphinx directive, so that it wouldn't be parsed in stuff like example output, but I don't know if that is possible to do, since it isn't the usual Sphinx directive syntax.

**Comment by asmeurer** [comment]
OK this is ready to review now.

**Comment by oscarbenjamin** [comment]
I'm not sure I understand the code well enough to comment but the general idea looks good. The dollars seem much more natural to me.

Will a PR fail the tests if it has some invalid markup like a single dollar symbol in a docstring?

**Comment by asmeurer** [comment]
No it won't actually. And it won't fail if it incorrectly replaces dollars in a doctest. I don't know if there is a way to make this more robust.

File: doc/ext/math_dollar.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    _data = {}
+    def repl(matchobj):
+        global _data
+        s = matchobj.group(0)
+        t = "___XXX_REPL_%d___" % len(_data)
+        _data[t] = s
+        return t
+    s = re.sub(r"({[^{}$]*\$[^{}$]*\$[^{}]*})", repl, s)
+    # matches $...$
+    dollars = re.compile(r"(?<!\$)(?<!\\)\$([^\$ ][^\$]*?)\$")
```
Comment (jksuom) [inline doc/ext/math_dollar.py]: Could these compiled re's be global?

Comment (oscarbenjamin) [reply doc/ext/math_dollar.py]: `re.compile` uses an internal cache:
https://docs.python.org/3/library/re.html#re.compile

**Comment by asmeurer** [comment]
Is there a way to check if something has `.. math` that isn't recognized by Sphinx as a math directive?

**Comment by asmeurer** [comment]
Maybe it's possible to hook directly into the doctree nodes, so that we only replace dollars in the proper places https://www.sphinx-doc.org/en/master/extdev/nodes.html

**Comment by asmeurer** [comment]
I think we can use https://www.sphinx-doc.org/en/master/extdev/appapi.html#event-doctree-read to modify the doctree. We just need to figure out which doctree nodes should be modified. I can't find any good documentation on it, so it will probably need to be done by trial and error.

**Comment by jksuom** [comment]
It seems that MathJax can handle double dollars for display math by default. It is only necessary to escape backslashes to make LaTeX work. It looks like this could be one implementation.
```
--- a/doc/ext/math_dollar.py
+++ b/doc/ext/math_dollar.py
@@ -39,10 +39,15 @@ def repl(matchobj):
     s = re.sub(r"({[^{}$]*\$[^{}$]*\$[^{}]*})", repl, s)
     # matches $...$
     dollars = re.compile(r"(?<!\$)(?<!\\)\$([^\$ ][^\$]*?)\$")
+    # matches $$...$$
+    doubledollars = re.compile(r"\$\$([^\$]+)\$\$")
     # regular expression for \$
     slashdollar = re.compile(r"\\\$")
     s = dollars.sub(r":math:`\1`", s)
     s = slashdollar.sub(r"$", s)
+    def escape_slashes(matchobj):
+        return matchobj.group(0).replace("\\", "\\\\")
+    s = doubledollars.sub(escape_slashes, s)
     # change the original {...} things in:
     for r in _data:
         s = s.replace(r, _data[r])
```

**Comment by asmeurer** [comment]
A lot of people from other projects have told me that they would find this useful, so I'm going to develop it as a separate Sphinx extension https://github.com/sympy/sphinx-math-dollar. I'll update this PR once I have the extension up and running.

**Comment by asmeurer** [comment]
I have a basic thing working. I still need to document it and package it up to actually enable it in this PR.

It turns out that this is harder to do "correctly" than you would think, because you need to be able to parse the $$ at just the right time, or else the normal RST parsing will mess it up. If you parse too late, then RST strips the escapes. If you parse too soon, you will grab things that are in other kinds of blocks like docstrings, or you will bypass some special replacements that Sphinx does like replacing quotes with curly quotes. I think the way I've done it now is correct. It's at least correct for the Bessel and LaTeX printer documentation, which I checked. 

I also had to fix the regular expressions that were used because they didn't handle escaped dollar signs correctly (like `$\$13$`). 

I've opened some issues on the Sphinx extension issue tracker for what remains to be done. https://github.com/sympy/sphinx-math-dollar/issues. If someone wants to help out with one of them, please let me know.

**Comment by asmeurer** [comment]
OK, I've finished and published the Sphinx extension, and enabled it here. This is ready for review.

**Comment by oscarbenjamin** [comment]
Should the sphinx-math-dollar extension be added to the installation instructions for building the docs?

**Comment by asmeurer** [comment]
Oh yeah I forgot about that. Let me add it.

**Comment by oscarbenjamin** [comment]
This looks good to me.

Perhaps a different issue but the instructions for building the docs in the main README are inadequate:
https://github.com/sympy/sympy/blob/master/README.rst#documentation-and-usage

Maybe that should just say to look at doc/README

**Comment by asmeurer** [comment]
I didn't know they were there too. @lglattly that's something to make a note of. We discussed today at the GSoD meeting that we are going to move the doc build instructions in to the documentation guide that @lglattly is writing, and change the existing references (README, doc/README, the wiki page) to point to that guide.

**Comment by lglattly** [comment]
I've made note of this additional location for instructions on building the docs. I think it would be very useful to streamline all of the different sources for writing documentation into the one guide I am working on, so thanks for the heads up about this @oscarbenjamin!

**Comment by cbm755** [comment]
I've played with this.  I like it.  I'm +1 to merge.

**Pull request merged**